controller/volume_controller.go

```diff
@@ -1125,77 +1182,401 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
+		// reattach volume if detached unexpected and there are still healthy replicas
+		if e.Status.CurrentState == longhorn.InstanceStateError && v.Status.CurrentNodeID != "" {
+			log.Warn("Engine of volume dead unexpectedly, reattach the volume")
+			msg := fmt.Sprintf("Engine of volume %v dead unexpectedly, reattach the volume", v.Name)
+			vc.eventRecorder.Event(v, v1.EventTypeWarning, constant.EventReasonDetachedUnexpectly, msg)
+			e.Spec.LogRequested = true
+			for _, r := range rs {
+				if r.Status.CurrentState == longhorn.InstanceStateRunning {
+					r.Spec.LogRequested = true
+					rs[r.Name] = r
+				}
+			}
+			v.Status.Robustness = longhorn.VolumeRobustnessFaulted
+		}
 	}
 
-	if e == nil {
+	if err := vc.reconcileAttachDetachStateMachine(v, e, rs, isNewVolume, log); err != nil {
+		return err
+	}
+
+	if v.Status.CurrentNodeID != "" &&
+		v.Status.State == longhorn.VolumeStateAttached &&
+		e.Status.CurrentState == longhorn.InstanceStateRunning {
+		if e.Spec.RequestedBackupRestore != "" {
+			v.Status.Conditions = types.SetCondition(v.Status.Conditions,
+				longhorn.VolumeConditionTypeRestore, longhorn.ConditionStatusTrue, longhorn.VolumeConditionReasonRestoreInProgress, "")
+		}
+
+		// TODO: reconcileVolumeSize
+		// The engine expansion is complete
+		if v.Status.ExpansionRequired && v.Spec.Size == e.Status.CurrentSize {
+			v.Status.ExpansionRequired = false
+			v.Status.FrontendDisabled = false
+		}
+	}
+
+	if err := vc.checkAndFinishVolumeRestore(v, e, rs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) reconcileAttachDetachStateMachine(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica, isNewVolume bool, log *logrus.Entry) error {
+	//TODO: link the state machine graph here
+
+	if isNewVolume || v.Status.State == "" {
+		v.Status.State = longhorn.VolumeStateCreating
+		return nil
+	}
+
+	if v.Spec.NodeID == "" {
+		if v.Status.CurrentNodeID == "" {
+			switch v.Status.State {
+			case longhorn.VolumeStateAttached, longhorn.VolumeStateAttaching, longhorn.VolumeStateCreating:
+				vc.closeVolumeDependentResources(v, e, rs)
+				v.Status.State = longhorn.VolumeStateDetaching
+			case longhorn.VolumeStateDetaching:
+				vc.closeVolumeDependentResources(v, e, rs)
+				if vc.verifyVolumeDependentResourcesClosed(e, rs) {
+					v.Status.State = longhorn.VolumeStateDetached
+					vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonDetached, "volume %v has been detached", v.Name)
+				}
+			case longhorn.VolumeStateDetached:
+				// This is a stable state.
+				// We attempt to close the resources anyway to make sure that they are closed
+				vc.closeVolumeDependentResources(v, e, rs)
+			}
+			return nil
+		}
+		if v.Status.CurrentNodeID != "" {
+			switch v.Status.State {
+			case longhorn.VolumeStateAttaching, longhorn.VolumeStateAttached, longhorn.VolumeStateDetached:
```

Comment:
Curious. If `v.Status.State` is already `VolumeStateDetached`, is it possible that the `v.Status.CurrentNodeID` is not empty and we need to change it to `longhorn.VolumeStateDetaching` here?

Reply:
Agree that this state might not be possible. There are several more states that might not be possible but we include them in [the state machine](https://github.com/longhorn/longhorn/pull/4820/files#diff-fae42032720f1b29f136db6a03081662353b0ad870751a075eb5fb85ae63d06b) anyways so that it is guaranteed that our controller never gets into a deadlock. What do you think?